google/search/src/session.ts

```diff
@@ -34,117 +35,163 @@ async function clearAppCacheDir (): Promise<void> {
-    session.browserContext = await newBrowserContext(session.sessionDir)
-    return session
+  public async withSession<T>(sessionId: string, fn: (ctx: BrowserContext) => Promise<T>): Promise<T> {
+    const session = await this.acquireSession(sessionId)
+    try {
+      console.info(`[SessionManager] Running work on session ${session.sessionId}.`)
+      return await fn(session.browserContext)
+    } catch (err) {
+      console.error(`Error during work on session ${session.sessionId}: ${err}`)
+      throw err
+    } finally {
+      await this.releaseSession(sessionId)
+    }
   }
 
-  async close (): Promise<void> {
-    await this.browserContext?.close()
-    await fs.rm(this.sessionDir, { recursive: true })
-  }
-}
+  public async destroy(): Promise<void> {
+    console.info('[SessionManager] Destroying all sessions...')
+    this.isDestroyed = true
+
+    const sessionIds = Array.from(this.sessionCache.keys())
+    for (const sessionId of sessionIds) {
+      await this.lock.acquire(sessionId, async () => {
+        const session = this.sessionCache.get(sessionId)
+        if (session) {
+          console.info(`[Session ${sessionId}] Closing session as part of destroy().`)
+          await this.destroySession(session, true)
+        }
+      })
+    }
 
-async function mkSessionDir (sessionId: string): Promise<string> {
-  const sessionDir = path.resolve(APP_CACHE_DIR, 'browser_sessions', sessionId)
-  await fs.mkdir(sessionDir, { recursive: true })
-  return sessionDir
+    this.sessionCache.clear()
+    SessionManager.instance = null
+    console.info('[SessionManager] Destroy complete.')
+  }
 }
 
-export function getSessionId (headers: IncomingHttpHeaders | undefined): string {
-  const workspaceId = getWorkspaceId(headers?.['x-gptscript-env'])
+export function getSessionId(headers: IncomingHttpHeaders): string {
+  const workspaceId = getGPTScriptEnv(headers, 'GPTSCRIPT_WORKSPACE_ID')
   if (workspaceId == null) throw new Error('No GPTScript workspace ID provided')
 
   return createHash('sha256').update(workspaceId).digest('hex').substring(0, 16)
 }
 
-function getWorkspaceId (envHeader: string | string[] | undefined): string | undefined {
+export function getWorkspaceId(headers: IncomingHttpHeaders): string | undefined {
+  return getGPTScriptEnv(headers, 'GPTSCRIPT_WORKSPACE_ID')
+}
+
+export function getGPTScriptEnv(headers: IncomingHttpHeaders, envKey: string): string | undefined {
+  const envHeader = headers?.['x-gptscript-env']
   const envArray = Array.isArray(envHeader) ? envHeader : [envHeader]
-  for (const env of envArray) {
-    if (env == null) {
-      continue
-    }
 
+  for (const env of envArray) {
+    if (env == null) continue
     for (const pair of env.split(',')) {
-      const [key, value] = pair.split('=')
-      if (key === 'GPTSCRIPT_WORKSPACE_ID') {
-        return value ?? process.env.GPTSCRIPT_WORKSPACE_ID
-      }
+      const [key, value] = pair.split('=').map((part) => part.trim())
```

Comment:
oh, is the space that were causing the problem?

Reply:
heh, yup